db/active_replicator_checkpointer.go

```diff
@@ -133,43 +126,63 @@ func (c *Checkpointer) CheckpointNow() {
 	}
 }
 
-func (c *Checkpointer) calculateCheckpointSeq() (seq string) {
-	base.TracefCtx(c.ctx, base.KeyReplicate, "checkpointer: calculateCheckpointSeq(%v, %v)", c.expectedSeqs, c.processedSeqs)
+// _updateCheckpointLists determines the highest checkpointable sequence, and trims the processedSeqs/expectedSeqs lists up to this point.
+func (c *Checkpointer) _updateCheckpointLists() (safeSeq string) {
+	base.TracefCtx(c.ctx, base.KeyReplicate, "checkpointer: _updateCheckpointLists(expectedSeqs: %v, procssedSeqs: %v)", c.expectedSeqs, c.processedSeqs)
 
-	if len(c.expectedSeqs) == 0 {
+	maxI := c._calculateSafeExpectedSeqsIdx()
+	if maxI == -1 {
 		// nothing to do
 		return ""
 	}
 
+	safeSeq = c.expectedSeqs[maxI]
+
+	// removes to-be checkpointed sequences from processedSeqs list
+	for i := 0; i < maxI; i++ {
+		removeSeq := c.expectedSeqs[i]
+		delete(c.processedSeqs, removeSeq)
+		base.TracefCtx(c.ctx, base.KeyReplicate, "checkpointer: _updateCheckpointLists removed seq %v from processedSeqs map %v", removeSeq, c.processedSeqs)
+	}
+
+	if len(c.expectedSeqs)-1 == maxI {
+		// received full set, empty expectedSeqs list
+		c.expectedSeqs = c.expectedSeqs[0:0]
+	} else {
+		// trim expectedSeqs list for partially received set
+		c.expectedSeqs = c.expectedSeqs[maxI+1:]
```

Comment:
The second approach - [maxI+1:] - also empties the slice if len(c.expectedSeqs)-1 == maxI.  Is using [0:0] more memory efficient?

Reply:
Intending to prevent out of bounds. I forgot you can safely re-slice from after the last element of the slice.